Publish new package throw error when uplink set to `https://registry.npmmirror.com`（cnpm registry）

The registry `https://registry.npmmirror.com` will return a 413 status code for a GET request containing a request body, which will cause an error when publishing a new package, resulting in a failure to publish.

On the other hand, a GET request should not contain a request body.

OK! Thanks so much! @juanpicado 

<!-- PLEASE READ THIS:
 - If you are not sure is a bug, OPEN a DISCUSSION, if is a legitimate bug, is easy to create a bug from a discussion.
 - Empty reports won't be considered and eventually be closed by a bot.
 - Include debugging notes will help to fix it faster, HOW TO: https://github.com/verdaccio/verdaccio/wiki/Debugging-Verdaccio
 - If you remove this template, ticket will be closed immediately.
 - No English perfect is required, use public translators if is need it, we will do our best to help you.
 - Extra bonus: The most complete this report is delivered, the faster you will get a response.
 - Extra bonus: include screenshots, logs (remove sensitive data).
 - If you are willing to fix it, there is a checkbox at the bottom.
-->

## Your Environment

  <!-- bug below the version 5.x will be closed, see SECURITY.md for more details -->

- **verdaccio version**: v5.20.1
- **node version** [v16.19.0]:
- **package manager**: [npm@8.19.3]
- **os**: [CentOS Linux 7]
- **platform**: [npm]

## Describe the bug

<!-- A clear and concise description of what the bug is. -->

The registry https://registry.npmmirror.com will return a 413 status code for a GET request containing a request body, which will cause an error when publishing a new package, resulting in a failure to publish.

On the other hand, a GET request should not contain a request body.

## To Reproduce

### 1. Update config file

```yaml
uplinks:
  npmjs:
    url: https://registry.npmjs.org
  cnpmjs:
    url: https://registry.npmmirror.com
```

```yaml
packages:
  "**":
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: cnpmjs
```

### 2. Publish a new package named `tech-verdaccio-test`

```bash
npm publish
```

### 3. Result

The command line will get stuck, and after a while there will be a packet timeout error.

```log
npm notice
npm notice 📦  tech-verdaccio-test@1.0.13
npm notice === Tarball Contents ===
npm notice 1B    index.js
npm notice 2.1kB log.yml
npm notice 216B  package.json
npm notice === Tarball Details ===
npm notice name:          tech-verdaccio-test
npm notice version:       1.0.13
npm notice filename:      tech-verdaccio-test-1.0.13.tgz
npm notice package size:  1.1 kB
npm notice unpacked size: 2.3 kB
npm notice shasum:        97e85b2e839d73d8e77e0aac03ecce298f0794ef
npm notice integrity:     sha512-G3h70gHtHDDSt[...]XW3PUFGNz3v2Q==
npm notice total files:   3
npm notice
npm notice Publishing to http://192.168.1.197:4873/
npm ERR! code E503
npm ERR! 503 Service Unavailable - PUT http://192.168.1.197:4873/tech-verdaccio-test - one of the uplinks is down, refuse to publish

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/fulei/.npm/_logs/2023-02-09T16_26_16_338Z-debug-0.log
```

## Expected behavior

Publish success

## Screenshots, server logs, package manager log

```bash
tail -f ~/.pm2/logs/verdaccio-out.log
```

```log
http --- 192.168.8.37 requested 'PUT /tech-verdaccio-test'
info --- auth/allow_action: access granted to: undefined
info --- nobody is allowed publish for tech-verdaccio-test
http --- 200, user: nobody(192.168.8.37), req: 'PUT /tech-verdaccio-test', bytes: 2301/0
info --- making request: 'GET https://registry.npmmirror.com/tech-verdaccio-test'
http --- 413, req: 'GET https://registry.npmmirror.com/tech-verdaccio-test' (streaming)
http --- 413, req: 'GET https://registry.npmmirror.com/tech-verdaccio-test', bytes: 4/0
```

<!-- If applicable, add screenshots to help explain your problem.  -->

## Configuration File (cat ~/.config/verdaccio/config.yaml)

Refer to `To Reproduce`

<!-- Please be careful do not leak any sensitive information, remove tokens -->

## Environment information

```log
Environment Info:
  System:
    OS: Linux 3.10 CentOS Linux 7 (Core)
    CPU: (8) x64 Intel(R) Xeon(R) CPU           E5606  @ 2.13GHz
  Binaries:
    npm: 8.19.3 - ~/.nvm/versions/node/v16.19.0/bin/npm
  npmGlobalPackages:
    verdaccio: 5.20.1
```

## Contribute to Verdaccio

- [x] I'm willing to fix this bug if I have time 🥇
<!--

IMPORTANT: please do not attach external files, all content should be visible from any device.
-->

## How to fix

[./src/lib/up-storage.ts line 444](https://github.com/verdaccio/verdaccio/blob/bae430fe24c9bb63dcf7b7b9e5f9f0dbf3f42669/src/lib/up-storage.ts#L224)

```javascript
/**
   * Fetch an asset.
   * @param {*} options
   * @param {*} cb
   * @return {Request}
   */
  private request(options: any, cb?: Callback): Stream.Readable {
    // other codes

    let requestOptions = {
      url: uri,
      method: method,
      headers: headers,
      // This line causes the GET request to contain the request body
      // This line should be fixed as:
      // body: method === 'GET' ? undefined : json,
      body: json,
      proxy: this.proxy,
      encoding: null,
      gzip: true,
      timeout: this.timeout,
      strictSSL: this.strict_ssl,
      agentOptions: this.agent_options,
    };

    // other codes
  }
```

--- a/.changeset/poor-suns-film.md
+++ b/.changeset/poor-suns-film.md
@@ -1,0 +1,5 @@
+---
+'@verdaccio/core': patch
+---
+
+fix(core): fix `isObject` function.`isObject(true)` should return false.

--- a/packages/core/core/src/validation-utils.ts
+++ b/packages/core/core/src/validation-utils.ts
@@ -95,14 +95,15 @@ export function normalizeMetadata(manifest: Manifest, name: string): Manifest {
  * @return {Boolean}
  */
 export function isObject(obj: any): boolean {
-  if (obj === null || typeof obj === 'undefined' || typeof obj === 'string') {
-    return false;
-  }
+  // if (obj === null || typeof obj === 'undefined' || typeof obj === 'string') {
+  //   return false;
+  // }
 
-  return (
-    (typeof obj === 'object' || typeof obj.prototype === 'undefined') &&
-    Array.isArray(obj) === false
-  );
+  // return (
+  //   (typeof obj === 'object' || typeof obj.prototype === 'undefined') &&
+  //   Array.isArray(obj) === false
+  // );
+  return Object.prototype.toString.call(obj) === '[object Object]';
 }
 
 export function validatePassword(

--- a/packages/core/core/test/validation-utilts.spec.ts
+++ b/packages/core/core/test/validation-utilts.spec.ts
@@ -31,6 +31,7 @@ describe('isObject', () => {
     expect(isObject(['foo'])).toBeFalsy();
     expect(isObject(null)).toBeFalsy();
     expect(isObject(undefined)).toBeFalsy();
+    expect(isObject(true)).toBeFalsy();
   });
 });
 

